core/src/main/java/org/elasticsearch/common/network/NetworkModule.java

```diff
@@ -70,48 +71,76 @@
-     * @param networkService A constructed network service object to bind.
      * @param settings The settings for the node
      * @param transportClient True if only transport classes should be allowed to be registered, false otherwise.
      */
-    public NetworkModule(NetworkService networkService, Settings settings, boolean transportClient) {
-        this.networkService = networkService;
+    public NetworkModule(Settings settings, boolean transportClient) {
         this.settings = settings;
         this.transportClient = transportClient;
-        registerTransport(LOCAL_TRANSPORT, LocalTransport.class);
+        registerTransport(new NetworkPlugin.TransportFactory<Transport>(LOCAL_TRANSPORT) {
+
+            @Override
+            public Transport createTransport(Settings settings, ThreadPool threadPool, BigArrays bigArrays,
+                                             CircuitBreakerService circuitBreakerService, NamedWriteableRegistry namedWriteableRegistry,
+                                             NetworkService networkService) {
+                return new LocalTransport(settings, threadPool, namedWriteableRegistry, circuitBreakerService);
+            }
+        });
         namedWriteables.add(new NamedWriteableRegistry.Entry(Task.Status.class, ReplicationTask.Status.NAME, ReplicationTask.Status::new));
         namedWriteables.add(new NamedWriteableRegistry.Entry(Task.Status.class, RawTaskStatus.NAME, RawTaskStatus::new));
         registerBuiltinAllocationCommands();
     }
 
+    public void processPlugins(List<NetworkPlugin> plugins) {
+        for (NetworkPlugin plugin : plugins) {
+            if (transportClient == false && HTTP_ENABLED.get(settings)) {
+                List<NetworkPlugin.TransportFactory<HttpServerTransport>> httpTransportFactory = plugin.getHttpTransportFactory();
+                for (NetworkPlugin.TransportFactory<HttpServerTransport> factory : httpTransportFactory) {
+                    registerHttpTransport(factory);
+                }
+            }
+            List<NetworkPlugin.TransportFactory<Transport>> transportFactory = plugin.getTransportFactory();
+            for (NetworkPlugin.TransportFactory<Transport> factory : transportFactory) {
+                registerTransport(factory);
+            }
+            List<TransportInterceptor> transportInterceptors = plugin.getTransportInterceptors();
+            for (TransportInterceptor interceptor : transportInterceptors) {
+                registerTransportInterceptor(interceptor);
+            }
+        }
+    }
+
     public boolean isTransportClient() {
         return transportClient;
     }
 
     /** Adds a transport implementation that can be selected by setting {@link #TRANSPORT_TYPE_KEY}. */
-    public void registerTransport(String name, Class<? extends Transport> clazz) {
-        transportTypes.registerExtension(name, clazz);
+    public void registerTransport(NetworkPlugin.TransportFactory<Transport> factory) {
```

Comment:
Can these register methods now be private? Do we really need separate methods, ie the 2 lines here could be moved into the loop over plugin transports?